lib/resources/elasticsearch.rb

```diff
@@ -0,0 +1,168 @@
+          .add(:exists?) { |x| !x.entries.empty? }
+          .add(:count) { |t, _|
+            t.entries.length
+          }
+
+    filter.connect(self, :nodes)
+
+    attr_reader :nodes
+    def initialize(nodes)
+      @nodes = nodes
+    end
+  end
+
+  class Elasticsearch < Inspec.resource(1)
+    name 'elasticsearch'
+    desc "Use the Elasticsearch InSpec audit resource to test the Elasticsearch
+      configuration settings. The data is extract from the live elastic search
+      cluster using the curl command `curl localhost:9200/_nodes/`. The resource
+      makes uses of FilterTable and its functionality can be used "
+
+    example "
+      describe elasticsearch(es_ip:'0.0.0.0', es_port:'9200', es_user:'elastic', es_pass:'changeme', https:true, self_signed_cert:true).nodes do
+        it { should exist }
+      end
+
+      describe elasticsearch.nodes do
+        its('name') { should include 'my_node' }
+      end
+
+      describe elasticsearch.nodes do
+        its('name') { should_not include 'node_name' }
+      end
+
+      elasticsearch.nodes.os.each do |nodeos|
+        describe nodeos do
+          its('name') { should_not cmp 'MacOS' }
+        end
+      end
+
+      elasticsearch.nodes.version.each do |node_version|
+        describe node_version do
+          it { should be > '1.2.0' }
+        end
+      end
+
+      elasticsearch.nodes.where {os.name == 'Linux' }.settings.each do |node_settings|
+        describe node_settings do
+          its('path.conf') { should cmp '/etc/elasticsearch' }
+        end
+      end
+
+      elasticsearch.nodes.where {name == 'package-centos-72' }.settings.each do |node_settings|
+        describe node_settings do
+          its('path.conf') { should cmp '/etc/elasticsearch' }
+        end
+      end
+        "
+
+    def initialize(opts = {}) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      return skip_resource 'Package `curl` not avaiable on the host' unless inspec.command('curl').exist?
+
+      es_ip            = opts.fetch(:es_ip, '0.0.0.0')
+      es_port          = opts.fetch(:es_port, '9200')
+      es_user          = opts.fetch(:es_user, nil)
+      es_pass          = opts.fetch(:es_pass, nil)
+      https            = opts.fetch(:https, true)
+      self_signed_cert = opts.fetch(:self_signed_cert, false)
+
+      url = "-H 'Content-Type: application/json' "
+      url += https ? 'https' : 'http'
+      url += "://#{es_ip}:#{es_port}/_nodes/"
+      url += " -u #{es_user}:#{es_pass}" unless es_user.nil? || es_pass.nil?
+      url += ' -k' if self_signed_cert.eql?(true)
+
+      cmd = inspec.command("curl #{url}")
+      return skip_resource 'Connection refused please check ip and port provided' if cmd.stderr =~ /Failed to connect/
+      return skip_resource "Connection refused Peer's Certificate issuer is not recognized" if cmd.stderr =~ /Peer's Certificate issuer is not recognized/
```

Comment:
Is this necessary? The stderr will be displayed in the skip_resource message in line 112.

Reply:
This would help capture the reason of failure.

Reply:
Right, but the reason for the failure you're matching on ("Peer's Certificate issue is not recognized") is in the stderr output already, and it will be output in line 112. The skip message you're returning in this line 111 is not adding any additional value that wouldn't be handled properly in line 112.

Reply:
Oh i remmember, the Peer certificate stderr is this big ugly block
"% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\ncurl: (60) Peer's Certificate issuer is not recognized.\nMore details here: http://curl.haxx.se/docs/sslcerts.html\n\ncurl performs SSL certificate verification by default, using a \"bundle\"\n of Certificate Authority (CA) public keys (CA certs). If the default\n bundle file isn't adequate, you can specify an alternate file\n using the --cacert option.\nIf this HTTPS server uses a certificate signed by a CA represented in\n the bundle, the certificate verification probably failed due to a\n problem with the certificate (it might be expired, or the name might\n not match the domain name in the URL).\nIf you'd like to turn off curl's verification of the certificate, use\n the -k (or --insecure) option.\n"

I figured this would be a cleaner skip message of a known possible error.

Reply:
That's a great reason. Let's add a comment above this line that explains why we're specifically catching this one error type.

Reply:
Done!